### ISSUE TITLE
ListSerializer.to_representation

### ISSUE THREAD

**Issue by dilyanpalauzov** [encode/django-rest-framework#5790]
```python
def ListSerializer.to_representation:
        iterable = data.all() if isinstance(data, models.Manager) else data
        return [self.child.to_representation(item) for item in iterable]
```
A manager can have `.all()` method, but not be iterable.  E.g. `models.Manager` is defined as:
```python
class Manager(manager.BaseManager.from_queryset(query.QuerySet)):
    pass
```
and is not iterable.  Creating and using another manager:
```python
class Manager2(manager.BaseManager.from_queryset(newQuerySet)):
    pass
```
does not work with `ListSerializer.to_representation()` because it is not derived from `models.Manager` and is not iterable.

I propose changing the implementation to
```python
iterable = data.all() if hasattr(data, 'all') else data
```

**Comment by xordoquy** [comment]
We can't do much without context.
I just don't get your issue: `data.all()` will not return a manager. Why would it matter that it's not iterable then ?

**Comment by dilyanpalauzov** [comment]
The current code is
```
iterable = data.all() if hasattr(data, 'all') else data
return [self.child.to_representation(item) for item in iterable]
```
so `iterable` has to be an iterable, not necessary a Manager.

**Comment by xordoquy** [comment]
I don't see how iterable could be a manager. The call to `all` ensures it is an iterable because `QuerySet` are.
The issue isn't currently actionable. The next step for us to consider this issue would be to have a test case that demonstrate the issue so we can decide whether it's an issue with DRF or with the external code.

**Comment by dilyanpalauzov** [comment]
If `data` can be a manager, which is tested by `isinstance(data, models.Manager)` then apparently iterable can also be a manager, that is not an instance of `models.Manager`, as I originally mentioned.

The test case is: https://code.djangoproject.com/ticket/28668 adds a method to QuerySet.objects.bulk_create() that can send `INSERT INTO  .. ON CONFLICT DO NOTHING`.  I want to integrate this in my code now.  So I create a method `bulk_create2()` and do
```python
from django.db.models import manager, query

query.QuerySet.bulk_create2 = bulk_create2
class Manager2(manager.BaseManager.from_queryset(query.QuerySet)):
  pass
models.Manager = Manager2
```
and import this module in my models.py.  From then on the standard queryset knows the `bulk_create2` method.  However to `ListSerializers.to_representation` the passed `data` is not of the original type `models.Manager`, at the time serializers.py was loaded, but the Manager2, at the time `to_representation()` is executed.

To sum up, with a manager created in the aforementioned way, ListSerializers.to_representation() does not work.

**Comment by xordoquy** [comment]
Ensure you monkey patch django's model before drf is loaded and this is fixed.
monkey patching like this will probably also hit you elsewhere in some other lib.
A probably safer way would be to define a base model class for your project and ensure your models do inherit from it.

**Comment by dilyanpalauzov** [comment]
This test fails:
```python
from django.db import models


class Manager2(models.manager.BaseManager.from_queryset(models.query.QuerySet)):
    pass


models.Manager = Manager2
from django.test import TestCase
from rest_framework import serializers


class Model2(models.Model):
    z = models.IntegerField()


class TestListSerializerWithCustomManager(TestCase):
    def setUp(self):
        Model2.objects.create(z=1)

    def test_to_representation(self):
        class BasicModelSerializer2(serializers.ModelSerializer):
            class Meta:
                model = Model2
                fields = '__all__'

        class TestListSerializer2(serializers.ListSerializer):
            child = BasicModelSerializer2()

        serializer = TestListSerializer2()
        self.assertEqual(str(serializer.to_representation(Model2.objects)),
                         "[OrderedDict([('id', 1), ('z', 1)])]")
```

I guess it does not matter when DRF is loaded, as it can be seen from the imports.

**Comment by rpkilby** [comment]
@dilyanpalauzov - I believe #5782 addresses this issue.

**Comment by xordoquy** [comment]
Nice catch @rpkilby. It should indeed fix this issue.
